openhands/runtime/client/runtime.py

```diff
@@ -34,6 +35,62 @@
 from openhands.runtime.utils.runtime_build import build_runtime_image
 
 
+class LogBuffer:
+    """
+    Synchronous buffer for Docker container logs.
+
+    This class provides a thread-safe way to collect, store, and retrieve logs
+    from a Docker container. It uses a deque with a maximum length to store log
+    lines and provides methods for appending, retrieving, and clearing logs.
+    """
+
+    def __init__(self, container: docker.models.containers.Container):
+        self.buffer: list[str] = []
+        self.lock = threading.Lock()
+        self.log_generator = container.logs(stream=True, follow=True)
+        self.log_stream_thread = threading.Thread(target=self.stream_logs)
+        self.log_stream_thread.daemon = True
+        self.log_stream_thread.start()
+        self._stop_event = threading.Event()
+
+    def append(self, log_line: str):
+        with self.lock:
+            self.buffer.append(log_line)
+
+    def get_and_clear(self) -> list[str]:
+        with self.lock:
+            logs = list(self.buffer)
+            self.buffer.clear()
+            return logs
+
+    def stream_logs(self):
+        """
+        Stream logs from the Docker container in a separate thread.
+
+        This method runs in its own thread to handle the blocking
+        operation of reading log lines from the Docker SDK's synchronous generator.
+        """
+        try:
+            for log_line in self.log_generator:
+                if self._stop_event.is_set():
+                    break
+                if log_line:
+                    self.append(log_line.decode('utf-8').rstrip())
+        except Exception as e:
+            logger.error(f'Error in stream_logs: {e}')
+
+    def __del__(self):
+        if self.log_stream_thread.is_alive():
+            logger.warn(
+                "LogBuffer was not properly closed. Use 'log_buffer.close()' for clean shutdown."
+            )
+            self.close(timeout=5)
+
+    def close(self, timeout: float = 10.0):
+        self._stop_event.set()
+        self.log_stream_thread.join(timeout)
+
+
 class EventStreamRuntime(Runtime):
     """This runtime will subscribe the event stream.
     When receive an event, it will send the event to od-runtime-client which run inside the docker environment.
```

Comment:
`When it receives an event, it will send the event to od-runtime-client which runs inside the docker environment.`

Reply:
fixed!